From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sic <31908995+sicXnull@users.noreply.github.com>
Date: Tue, 20 Aug 2024 08:35:00 +0200
Subject: [ARCHEOLOGY] Add Board X96Q TV Box LPDDR3 H313 (#7101)

> X-Git-Archeology: > recovered message: > * X96-Q TV Box Support
> X-Git-Archeology: > recovered message: > * Update x96q-tvbox.csc
> X-Git-Archeology: > recovered message: > * Rename x96q-tvbox.csc to x96q.tvb
> X-Git-Archeology: - Revision 100a004aa5985b82184f1f00855461ee8f7a9f56: https://github.com/armbian/build/commit/100a004aa5985b82184f1f00855461ee8f7a9f56
> X-Git-Archeology:   Date: Tue, 20 Aug 2024 08:35:00 +0200
> X-Git-Archeology:   From: sic <31908995+sicXnull@users.noreply.github.com>
> X-Git-Archeology:   Subject: Add Board X96Q TV Box LPDDR3 H313 (#7101)
> X-Git-Archeology:
> X-Git-Archeology: - Revision 5e54e2beb8dc78c374e6ad741b86f7b01b466e1a: https://github.com/armbian/build/commit/5e54e2beb8dc78c374e6ad741b86f7b01b466e1a
> X-Git-Archeology:   Date: Wed, 18 Sep 2024 16:31:00 +0200
> X-Git-Archeology:   From: sicXnull <cvanhauterjr@gmail.com>
> X-Git-Archeology:   Subject: X96Q LPDDR3 Improvements
> X-Git-Archeology:
> X-Git-Archeology: - Revision 8bc31a28823328451506e7788dde967433ba7254: https://github.com/armbian/build/commit/8bc31a28823328451506e7788dde967433ba7254
> X-Git-Archeology:   Date: Thu, 19 Sep 2024 10:10:20 +0200
> X-Git-Archeology:   From: Igor Pecovnik <igor@armbian.com>
> X-Git-Archeology:   Subject: Bugfix: resolve boot loader compilation issues on X96-mate and Orangepizero2
> X-Git-Archeology:
---
 arch/arm/dts/Makefile                    |   1 +
 arch/arm/dts/sun50i-h313-x96q-lpddr3.dts | 162 ++++++++++
 configs/x96q_lpddr3_defconfig            |  31 ++
 3 files changed, 194 insertions(+)

diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 111111111111..222222222222 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -836,6 +836,7 @@ dtb-$(CONFIG_MACH_SUN50I_H6) += \
 dtb-$(CONFIG_MACH_SUN50I_H616) += \
 	sun50i-h616-orangepi-zero2.dtb \
 	sun50i-h618-orangepi-zero3.dtb \
+	sun50i-h313-x96q-lpddr3.dtb \
 	sun50i-h616-x96-mate.dtb
 dtb-$(CONFIG_MACH_SUN50I) += \
 	sun50i-a64-amarula-relic.dtb \
diff --git a/arch/arm/dts/sun50i-h313-x96q-lpddr3.dts b/arch/arm/dts/sun50i-h313-x96q-lpddr3.dts
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/arch/arm/dts/sun50i-h313-x96q-lpddr3.dts
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: (GPL-2.0+ or MIT)
+/*
+ * Author: piotr.oniszczuk@gmail.com
+ */
+
+/dts-v1/;
+
+#include "sun50i-h616.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	model = "hechuang,x96q LPDDR3";
+	compatible = "hechuang,x96-q", "allwinner,sun50i-h616";
+
+	aliases {
+		mmc0 = &mmc0;
+		mmc1 = &mmc1;
+		mmc2 = &mmc2;
+		ethernet0 = &emac1;
+		serial0 = &uart0;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led-red {
+			function = LED_FUNCTION_DISK_ACTIVITY;
+			color = <LED_COLOR_ID_RED>;
+			gpios = <&pio 7 6 GPIO_ACTIVE_HIGH>; /* PH6 */
+			linux,default-trigger = "mmc0";
+		};
+	};
+
+	reg_vcc5v: vcc5v {
+		/* board wide 5V supply directly from the USB-C socket */
+		compatible = "regulator-fixed";
+		regulator-name = "vcc-5v";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		regulator-always-on;
+	};
+};
+
+&cpu0 {
+	cpu-supply = <&reg_dcdc2>;
+	status = "okay";
+};
+
+&cpu1 {
+	cpu-supply = <&reg_dcdc2>;
+	status = "okay";
+};
+
+&cpu2 {
+	cpu-supply = <&reg_dcdc2>;
+	status = "okay";
+};
+
+&cpu3 {
+	cpu-supply = <&reg_dcdc2>;
+	status = "okay";
+};
+
+&emac1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&rmii_pins>;
+	phy-mode = "rmii";
+	phy-handle = <&rmii_phy>;
+	phy-supply = <&reg_aldo1>;
+	allwinner,rx-delay-ps = <3100>;
+	allwinner,tx-delay-ps = <700>;
+	status = "okay";
+};
+
+&mdio1 {
+	rmii_phy: ethernet-phy@0 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <0>;
+	};
+};
+
+&mmc0 {
+	vmmc-supply = <&reg_dldo2>;
+	broken-cd;
+	bus-width = <4>;
+	status = "okay";
+};
+
+&mmc1 {
+	vmmc-supply = <&reg_dldo2>;
+	broken-cd;
+	bus-width = <4>;
+	status = "okay";
+};
+
+&mmc2 {
+	vmmc-supply = <&reg_dldo2>;
+	bus-width = <8>;
+	non-removable;
+	cap-mmc-hw-reset;
+	status = "okay";
+};
+
+&r_i2c {
+	status = "okay";
+
+	axp313a: pmic@36 {
+		compatible = "x-powers,axp313a";
+		reg = <0x36>;
+		wakeup-source;
+
+		regulators {
+			reg_dcdc1: dcdc1 {
+				regulator-always-on;
+				regulator-min-microvolt = <1160000>;
+				regulator-max-microvolt = <1160000>;
+				regulator-name = "vdd-cpu";
+			};
+
+			reg_dcdc2: dcdc2 {
+				regulator-always-on;
+				regulator-min-microvolt = <1160000>;
+				regulator-max-microvolt = <1160000>;
+				regulator-name = "vdd-gpu-sys";
+			};
+
+			reg_dcdc3: dcdc3 {
+				regulator-always-on;
+				regulator-min-microvolt = <1200000>;
+				regulator-max-microvolt = <1200000>;
+				regulator-name = "vdd-dram";
+			};
+
+			reg_aldo1: ldo1 {
+				regulator-always-on;
+				regulator-min-microvolt = <1800000>;
+				regulator-max-microvolt = <1800000>;
+				regulator-name = "vcc-sys";
+			};
+
+			reg_dldo2: ldo2 {
+				regulator-always-on;
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-name = "vcc3v3-ext";
+			};
+		};
+	};
+};
+
+&uart0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart0_ph_pins>;
+	status = "okay";
+};
diff --git a/configs/x96q_lpddr3_defconfig b/configs/x96q_lpddr3_defconfig
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/configs/x96q_lpddr3_defconfig
@@ -0,0 +1,31 @@
+CONFIG_ARM=y
+CONFIG_ARCH_SUNXI=y
+CONFIG_DEFAULT_DEVICE_TREE="sun50i-h313-x96q-lpddr3"
+CONFIG_SPL=y
+CONFIG_SUNXI_DRAM_H616_LPDDR3=y
+CONFIG_DRAM_CLK=600
+CONFIG_DRAM_SUN50I_H616_DX_ODT=0x06060606
+CONFIG_DRAM_SUN50I_H616_DX_DRI=0x0d0d0d0d
+CONFIG_DRAM_SUN50I_H616_CA_DRI=0x00000d0d
+CONFIG_DRAM_SUN50I_H616_ODT_EN=0x00000001
+CONFIG_DRAM_SUN50I_H616_TPR0=0x0
+CONFIG_DRAM_SUN50I_H616_TPR2=0x00000000
+CONFIG_DRAM_SUN50I_H616_TPR10=0x002f3359
+CONFIG_DRAM_SUN50I_H616_TPR11=0xaa889967
+CONFIG_DRAM_SUN50I_H616_TPR12=0xeeee8979
+CONFIG_MACH_SUN50I_H616=y
+CONFIG_R_I2C_ENABLE=y
+CONFIG_SPL_I2C=y
+CONFIG_SPL_I2C_SUPPORT=y
+CONFIG_SPL_SYS_I2C_LEGACY=y
+CONFIG_SYS_I2C_MVTWSI=y
+CONFIG_SYS_I2C_SLAVE=0x7f
+CONFIG_SYS_I2C_SPEED=100000
+CONFIG_SYS_MONITOR_LEN=786432
+CONFIG_PHY_REALTEK=y
+CONFIG_SUN8I_EMAC=y
+CONFIG_I2C3_ENABLE=y
+CONFIG_AXP313_POWER=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_OHCI_HCD=y
+CONFIG_USB_MUSB_GADGET=y
-- 
Armbian

